### PR TITLE
Rework multi-transition support using a recursive composition API

### PR DESCRIPTION
**Pull request by jverkoey**
This largely reworks the changes made in #40.

This new API supports a recursive model for building modular transitions. Every view controller transition is now governed by a single root transition. This root transition can then compose out to child transitions if it so chooses. A view controller transition is complete once every transition in the tree is complete.

This pattern provides clear solutions for how to handle data passing and sharing between modular transitions, as seen most clearly in the `PhotoAlbumTransition` example which makes heavy use of transition composition and view duplication.

There are four major parts to this change:

1. Updated the README documentation.
2. Updated the examples.
3. Added the functionality.
4. Added more tests.

## README

The readme's intro tutorial on how to build a transition has been updated.

## Updated examples

The photo album transition example has been updated to make use of compositional transitions. Its transition hierarchy looks like so:

transitions

## Added functionality

The new compositional functionality is made visible to clients through TransitionContext's new API `compose(with:)`. For example:

```swift
context.compose(with: FadeTransition(target: .foreView, style: .fadeIn))
context.compose(with: SpringFrameTransition(target: .target(snapshotContextView),
                                            size: fitSize))
```

The implementation of the tree-based compositional structure is found in `MDMViewControllerTransitionCoordinator.m`.

## Added tests

There are new tests for validating certain features of the transitioning stack, including the new compositional APIs.

## Future plans

With this architecture in place, we'll be able to start deploying a library of pre-made transitions that clients can piece together to build complex transitions. These transitions will go in a separate MaterialMotion Objective-C library, making use of both MotionTransitioning and MotionAnimator.

### CODE REVIEW

File: src/MDMTransitionContext.h [hunk 1]
```diff
@@ -84,6 +86,15 @@ NS_SWIFT_NAME(TransitionContext)
 @property(nonatomic, strong, readonly, nullable) UIPresentationController *presentationController;
 
 /**
+ Adds the provided transition as a child of the current transition and invokes its start method.
+
+ Each child transition will receive its own transition context instance to which the transition must
+ eventually invoke transitionDidEnd. Only once both the parent transition and all of its children
+ (and their children) have completed will the overall view controller transition be completed.
+ */
+- (void)composeWithTransition:(nonnull id<MDMTransition>)transition;
```
Comment (jverkoey) [inline src/MDMTransitionContext.h:95]: Feedback requested on this API name.

File: src/MDMTransitionController.h [hunk 2]
```diff
@@ -28,26 +28,21 @@ NS_SWIFT_NAME(TransitionController)
...
+@property(nonatomic, strong, nullable) id<MDMTransition> transition;
 
 /**
- The active transition instances.
+ The active transition instance.
 
  This may be non-nil while a transition is active.
  */
-@property(nonatomic, strong, nullable, readonly) NSArray<id<MDMTransition>> *activeTransitions;
+@property(nonatomic, strong, nullable, readonly) id<MDMTransition> activeTransition;
```
Comment (jverkoey) [inline src/MDMTransitionController.h:46]: The changes here are mostly reverting the changes introduced in 8653958a5a9419891861fb6fd7648791ca3c744c.

File: src/MDMTransitionPresentationController.m [hunk 3]
```diff
@@ -113,6 +113,8 @@ - (void)startWithContext:(NSObject<MDMTransitionContext> *)context {
 
     [UIView animateWithDuration:context.duration animations:^{
       self.scrimView.alpha = context.direction == MDMTransitionDirectionForward ? 1 : 0;
+    } completion:^(BOOL finished) {
+      [context transitionDidEnd];
```
Comment (jverkoey) [inline src/MDMTransitionPresentationController.m:117]: Presentation controllers also need to inform the context of completion.

File: src/private/MDMViewControllerTransitionCoordinator.m [hunk 4]
```diff
@@ -18,18 +18,25 @@
...
-@property(nonatomic, strong) id<MDMTransition> transition;
+@class MDMViewControllerTransitionContextNode;
+
+@protocol MDMViewControllerTransitionContextNodeParent <NSObject>
+- (void)childNodeTransitionDidEnd:(MDMViewControllerTransitionContextNode *)childNode;
 @end
 
-@protocol MDMViewControllerTransitionContextDelegate <NSObject>
-- (void)transitionContextDidEnd:(MDMViewControllerTransitionContext *)context;
+@interface MDMViewControllerTransitionContextNode : NSObject <MDMTransitionContext, MDMViewControllerTransitionContextNodeParent>
```
Comment (jverkoey) [inline src/private/MDMViewControllerTransitionCoordinator.m:27]: Typical parent/node design pattern here. The parent protocol only exposes APIs needed by the child to communicate upwards.

File: src/private/MDMViewControllerTransitionCoordinator.m [hunk 5]
```diff
@@ -18,18 +18,25 @@
...
+@property(nonatomic, strong) id<UIViewControllerContextTransitioning> transitionContext;
+@property(nonatomic, strong, readonly) id<MDMTransition> transition;
+@property(nonatomic, copy, readonly) NSMutableArray<MDMViewControllerTransitionContextNode *> *children;
 @end
 
-@implementation MDMViewControllerTransitionContext {
+@implementation MDMViewControllerTransitionContextNode {
+  // Every node points to the same array in memory.
```
Comment (jverkoey) [inline src/private/MDMViewControllerTransitionCoordinator.m:34]: This is a minor simplification: by sharing the same mutable array, every node can add to a single shared pool of completion blocks. The alternative would be to route messages up the parent hierarchy, but given that this is a private implementation detail I wasn't particularly concerned with such an approach.

File: src/private/MDMViewControllerTransitionCoordinator.m [hunk 6]
```diff
@@ -46,21 +53,135 @@ - (instancetype)initWithTransition:(id<MDMTransition>)transition
...
+  BOOL anyChildActive = NO;
+  for (MDMViewControllerTransitionContextNode *child in _children) {
+    if (!child->_didEnd) {
+      anyChildActive = YES;
+      break;
+    }
+  }
+
+  if (!anyChildActive && _didEnd) { // Inform our parent of completion.
+    [_parent childNodeTransitionDidEnd:self];
```
Comment (jverkoey) [inline src/private/MDMViewControllerTransitionCoordinator.m:98]: Bubbles completion up the tree.

**Comment by jverkoey** [comment]
Hi reviewers, I've left some comments to help guide review of the source.

**Comment by jverkoey** [comment]
Gentle ping.

**Review by ajsecord — APPROVED** [review]
And Lo! In the days of [REDACTED], there was a strange silence upon the land, wherein the voices of many were missing or muted. Houses went unbuilt, promises were not kept, and there was a general uneasiness about all.

**Pull request merged**